Fixed #30907 -- Fixed SplitArrayField.has_changed() with removal of empty trailing values.

paveldedik opened the pull request:
https://code.djangoproject.com/ticket/30907

`django.contrib.postgres.forms.SplitArrayField.has_changed()` returns `True` when `initial` is empty list and `data` is list of empty strings. This is a problem when a form is submitted with input fields left blank - resulting in e.g. `["", "",]` (passed to `has_changed()` and evaluated as `True` instead of `False`). I believe the issue I'm describing should be more clear from the written tests.

Note that this bug in `has_changed()` method causes issues with FormSets which fail to correctly validate empty forms (in cases where `SplitArrayField` is used).

The issue is similar to https://github.com/django/django/pull/9520

paveldedik commented:
I updated the commit (check the diff) as I realized the behavior I implemented only makes sense when using `remove_trailing_nulls=True`. See https://docs.djangoproject.com/en/2.2/ref/contrib/postgres/forms/#django.contrib.postgres.forms.SplitArrayField.remove_trailing_nulls

I'm not sure what makes sense in case `remove_trailing_nulls=False`, probably the default behavior.

felixxm commented on django/contrib/postgres/forms/array.py:
```diff
@@ -216,3 +216,23 @@ def clean(self, value):
+
+        if initial in self.empty_values and data in self.empty_values:
+            return False
+        return super().has_changed(initial, data)
```
I think we can add an internal hook `_remove_trailing_nulls()` to the `SplitArrayField` that can be used in `clean()` and `has_changed()`:
```python
    def _remove_trailing_nulls(self, values):
        errors = []
        null_index = None
        if self.remove_trailing_nulls:
            for i, value in reversed(list(enumerate(values))):
                if value in self.base_field.empty_values:
                    null_index = i
                else:
                    break
            if null_index is not None:
                values = values[:null_index]
        return values, null_index

    def has_changed(self, initial, data):
        data, _ = self._remove_trailing_nulls(self.to_python(data))
        return super().has_changed(initial, data)
```
Please add it in the 1st commit, the 2nd commit will contain a bugfix.

paveldedik replied:
This on its own will not fix the issue if `initial` is `None`. We also need to add:

```python
if initial is in self.empty_values and data is in self.empty_values:
   return False
```

Or something similar. At least I think, the assertion: `self.assertIs(field.has_changed(None, ['', '']), False)` is covering this case.

felixxm replied:
Yes, we should use the same approach as in `SimpleArrayField`, e.g.
```python
    def has_changed(self, initial, data):
        try:
            data, _ = self._remove_trailing_nulls(self.to_python(data))
        except ValidationError:
            pass
        if initial in self.empty_values and data in self.empty_values:
            return False            
        return super().has_changed(initial, data)
```

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -965,6 +965,37 @@ class Meta:
+        self.assertIs(field.has_changed([None], ['url', None]), True)
+        self.assertIs(field.has_changed(['a'], ['b', None]), True)
+
+    def test_splitarrayfield_has_changed_empty(self):
```
`test_splitarrayfield_has_changed_empty` -> 
`test_remove_trailing_nulls_has_changed`

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -965,6 +965,37 @@ class Meta:
+        self.assertFalse(form.has_changed())
+
+        field = SplitArrayField(forms.CharField(), required=False, size=2, remove_trailing_nulls=True)
+        self.assertIs(field.has_changed([], [None]), False)
```
Do we need all these tests? IMO these two should be enough:
```python
obj = IntegerArrayModel(field=[])
form = Form({'field_0': '', 'field_1': ''}, instance=obj)
self.assertFalse(form.has_changed())

obj = IntegerArrayModel(field=[1])
form = Form({'field_0': '1', 'field_1': ''}, instance=obj)
self.assertFalse(form.has_changed())
```

paveldedik replied:
I think `obj = IntegerArrayModel(field=None)` or `obj = IntegerArrayModel()` needs to be tested as well though.

felixxm replied:
Agreed.

felixxm commented on tests/postgres_tests/test_array.py:
```diff
@@ -965,6 +965,37 @@ class Meta:
+        self.assertIs(field.has_changed(['a'], ['a', '']), False)
+        self.assertIs(field.has_changed(['url'], ['', 'url']), True)
+        self.assertIs(field.has_changed([None], ['url', None]), True)
+        self.assertIs(field.has_changed(['a'], ['b', None]), True)
```
These assertions are not related with a bugfix, please move them to a separate commit.

paveldedik commented on django/contrib/postgres/forms/array.py:
```diff
@@ -216,3 +216,23 @@ def clean(self, value):
+        except ValidationError:
+            pass
+        else:
+            if self.remove_trailing_nulls:
```
But that would change the SplitArrayField interface breaking it for users...

pope1ni replied:
Sorry, I misinterpreted it as a new flag. I also see above that this was all copied so ignore this.

While not the best naming, we're stuck with it.

paveldedik commented on django/contrib/postgres/forms/array.py:
```diff
@@ -216,3 +216,23 @@ def clean(self, value):
+        else:
+            if self.remove_trailing_nulls:
+                null_index = None
+                for i, value in reversed(list(enumerate(python_data))):
```
Yep, you are right.

paveldedik replied:
Actually, it is required - `TypeError: 'enumerate' object is not reversible`. So I kept it there.

pope1ni replied:
That actually makes sense. I forgot `reversed()` is a bit special because you typically have to have consumed the whole iterator to know the last element.

paveldedik commented:
Fixed. Also note https://github.com/django/django/pull/11970/commits/5167db50ba48113d010944a7cfde5ab8b433afa7

I added assertions how the SplitArrayField behaves in case `remove_trailing_nulls=False`. Which is kind of weird behaviour, not sure it's correctly handled.

charettes commented on django/contrib/postgres/forms/array.py:
```diff
@@ -202,17 +214,19 @@ def clean(self, value):
+        try:
+            data, _ = self._remove_trailing_nulls(self.to_python(data))
+        except ValidationError:
+            pass
```
Since `self._remove_trailing_nulls` is not meant to `raise ValidationError` I'd only call `to_python` in the `try` block

```suggestion
        try:
            data = self.to_python(data)
        except ValidationError:
            pass
        else:
            data = self._remove_trailing_nulls(data)[0]
```

paveldedik replied:
Makes sense. Fixed

felixxm commented:
@paveldedik Thanks again :+1: I pushed minor edits and added tests for validation error in `has_changed()`.

Welcome aboard :boat:
